django/core/handlers/asgi.py

```diff
@@ -0,0 +1,342 @@
+import asyncio
+import cgi
+import codecs
+import logging
+import sys
+import traceback
+from io import BytesIO
+
+from asgiref.sync import sync_to_async
+
+from django import http
+from django.conf import settings
+from django.core import signals
+from django.core.exceptions import (
+    RequestAborted, RequestDataTooBig, RequestTimeout,
+)
+from django.core.handlers import base
+from django.http import FileResponse, HttpResponse, HttpResponseServerError
+from django.urls import set_script_prefix
+from django.utils.functional import cached_property
+
+logger = logging.getLogger("django.request")
+
+
+class AsgiRequest(http.HttpRequest):
+    """
+    Custom request subclass that decodes from an ASGI-standard request
+    dict, and wraps request body handling.
+    """
+
+    # Number of seconds until a Request gives up on trying to read a request
+    # body and aborts.
+    body_receive_timeout = 60
+
+    def __init__(self, scope, body):
+        self.scope = scope
+        self._content_length = 0
+        self._post_parse_error = False
+        self._read_started = False
+        self.resolver_match = None
+        self.script_name = self.scope.get("root_path", "")
+        if self.script_name and scope["path"].startswith(self.script_name):
+            # TODO: Better is-prefix checking, slash handling?
+            self.path_info = scope["path"][len(self.script_name):]
+        else:
+            self.path_info = scope["path"]
+
+        # django path is different from asgi scope path args, it should combine with script name
+        if self.script_name:
+            self.path = "%s/%s" % (
+                self.script_name.rstrip("/"),
+                self.path_info.replace("/", "", 1),
+            )
+        else:
+            self.path = scope["path"]
+
+        # HTTP basics
+        self.method = self.scope["method"].upper()
+        # Ensure query string is encoded correctly
+        query_string = self.scope.get("query_string", "")
+        if isinstance(query_string, bytes):
+            query_string = query_string.decode("utf-8")
+        self.META = {
+            "REQUEST_METHOD": self.method,
+            "QUERY_STRING": query_string,
+            "SCRIPT_NAME": self.script_name,
+            "PATH_INFO": self.path_info,
+            # WSGI-expecting code will need these for a while
+            "wsgi.multithread": True,
+            "wsgi.multiprocess": True,
+        }
+        if self.scope.get("client", None):
```

Comment:
`scope` is a dict, isn't the `None` already the default anyways? Same below.

Reply:
I often just put `None` in my `gets` just to make it clearer to everyone reading the code. I'll remove it.